File: backend/src/main/java/com/freight/backend/gpsload/routeassembly/model/RouteAssemblyRequest.java
package com.freight.backend.gpsmiss.routeassembly.model;

import java.util.List;

/**
 * 노선 조립 요청
 * 기사가 추천을 요청할 때 보내는 정보
 */
public record RouteAssemblyRequest(
        DriverState driverState,       // 기사 상태
        List<Quote> candidateQuotes,   // 후보 견적 목록 (선택적)
        AssemblyParameters parameters, // 알고리즘 파라미터 (선택적)
        List<Long> selectedQuoteIds,   // 기사가 미리 선택한 견적 ID (선택적)
        RouteMode mode                 // 노선 모드 (SIMPLE / SMART, 선택적)
) {
    /**
     * 노선 모드 열거형 — HTML 프론트엔드 이전
     * SIMPLE: 순차 왕복 (현재→상차1→하차1→상차2→하차2→...→복귀)
     * SMART : LIFO 알뜰배달 (현재→상차1→상차2→...→하차N→...→하차1→복귀)
     */
    public enum RouteMode {
        SIMPLE,
        SMART
    }

    /**
     * 파라미터가 없으면 기본값 사용
     */
    public AssemblyParameters getEffectiveParameters() {
        return parameters != null ? parameters : AssemblyParameters.defaults();
    }

    /**
     * 유효한 모드 반환 (null이면 SMART 기본)
     */
    public RouteMode getEffectiveMode() {
        return mode != null ? mode : RouteMode.SMART;
    }

    /**
     * 유효성 검증
     */
    public boolean isValid() {
        if (driverState == null) return false;
        if (driverState.currentLocation() == null) return false;
        if (driverState.currentLocation().latitude() == null) return false;
        if (driverState.currentLocation().longitude() == null) return false;
        return true;
    }

    /**
     * 후보 견적이 제공되었는지 확인
     */
    public boolean hasCandidateQuotes() {
        return candidateQuotes != null && !candidateQuotes.isEmpty();
    }

    /**
     * 선택된 견적이 있는지 확인
     */
    public boolean hasSelectedQuotes() {
        return selectedQuoteIds != null && !selectedQuoteIds.isEmpty();
    }
}

File: frontend/rodia-monorepo/apps/mobile/src/features/driver-reco/model/routeAssemblyRequest.ts
// OpenAPI grounding:
// - Quote { quoteId, origin: Place, destination: Place, volumeCbm, weightKg, allowCombine, combineAllowed,
//           finalPrice, scheduledDate, lengthCm, widthCm, heightCm, rotatable, stackable, fragile,
//           noStack, bottomOnly, maxStackWeight, status, handling }
// - Place { name?, address?, latitude?, longitude? }
// - RouteAssemblyRequest { selectedQuoteIds?, candidateQuotes?: Quote[], mode?: "SIMPLE"|"SMART" }
// - candidateQuotes with origin.latitude/longitude is required at runtime (server 500 without it)

import type { QuoteDetailResponse } from "@/entities/quote/model/quote.types";
import type { Quote, RouteAssemblyRequest } from "@/shared/api/generated/schemas";

function toOptionalFinite(v: unknown): number | undefined {
  const n = Number(v);
  return Number.isFinite(n) ? n : undefined;
}

function toOptionalStr(v: unknown): string | undefined {
  return typeof v === "string" && v.trim() ? v.trim() : undefined;
}

function toFinite(v: unknown, fallback?: number): number | undefined {
  const n = Number(v);
  if (Number.isFinite(n)) return n;
  return fallback;
}

/** Maps QuoteDetailResponse → generated Quote schema type (schema-grounded fields only). */
export function buildCandidateQuote(q: QuoteDetailResponse): Quote {
  return {
    quoteId: q.quoteId,
    origin: {
      name: toOptionalStr(q.originAddress),
      address: toOptionalStr(q.originAddress),
      latitude: toOptionalFinite(q.originLat),
      longitude: toOptionalFinite(q.originLng),
    },
    destination: {
      name: toOptionalStr(q.destinationAddress),
      address: toOptionalStr(q.destinationAddress),
      latitude: toOptionalFinite(q.destinationLat),
      longitude: toOptionalFinite(q.destinationLng),
    },
    volumeCbm: toOptionalFinite(q.volumeCbm),
    weightKg: toOptionalFinite(q.weightKg),
    allowCombine: q.allowCombine,
    combineAllowed: q.allowCombine,
    finalPrice: toOptionalFinite(q.finalPrice),
    status: q.status,
  };
}

/** Builds a RouteAssemblyRequest grounded in schema.
 *  Populates candidateQuotes so the server can resolve route geometry. */
export function buildRouteAssemblyRequest(params: {
  selectedQuoteIds: number[];
  quotes: QuoteDetailResponse[];
}): RouteAssemblyRequest {
  const { selectedQuoteIds, quotes } = params;
  const selectedIdSet = new Set(selectedQuoteIds);
  const candidateQuotes = quotes
    .filter((q) => selectedIdSet.has(q.quoteId))
    .map(buildCandidateQuote);

  const firstQuote = quotes[0];
  const lastQuote = quotes[quotes.length - 1];
  const currentLat = toFinite(firstQuote?.originLat, 37.5665);
  const currentLng = toFinite(firstQuote?.originLng, 126.978);
  const endLat = toFinite(lastQuote?.destinationLat);
  const endLng = toFinite(lastQuote?.destinationLng);
  const truckId = Number(firstQuote?.truckId);

  const routeMode: RouteAssemblyRequest["mode"] =
    selectedQuoteIds.length > 1 ? "SMART" : "SIMPLE";

  return {
    selectedQuoteIds,
    mode: routeMode,
    driverState: {
      currentLocation: {
        name: toOptionalStr(firstQuote?.originAddress),
        address: toOptionalStr(firstQuote?.originAddress),
        latitude: currentLat,
        longitude: currentLng,
      },
      ...(Number.isFinite(endLat) && Number.isFinite(endLng)
        ? {
            endLocation: {
              name: toOptionalStr(lastQuote?.destinationAddress),
              address: toOptionalStr(lastQuote?.destinationAddress),
              latitude: endLat,
              longitude: endLng,
            },
          }
        : {}),
      ...(Number.isInteger(truckId) && truckId > 0 ? { truckId } : {}),
    },
    ...(candidateQuotes.length > 0 ? { candidateQuotes } : {}),
  };
}
